proxy/proxy.go

```diff
@@ -0,0 +1,35 @@
+package proxy
+
+import (
+	"fmt"
+	"github.com/gin-gonic/gin"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+)
+
+type Controller struct {
+	prometheusUrl *url.URL
+}
+
+func NewController(prometheusUrl string) (*Controller, error) {
+	proxyUrl, err := url.Parse(prometheusUrl)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing proxy-url: %w", err)
+	}
+
+	return &Controller{prometheusUrl: proxyUrl}, nil
+}
+
+func (ctrl *Controller) Proxy(c *gin.Context) {
+	proxy := httputil.NewSingleHostReverseProxy(ctrl.prometheusUrl)
+	proxy.Director = func(req *http.Request) {
+		req.Header = c.Request.Header
+		req.Host = ctrl.prometheusUrl.Host
+		req.URL.Scheme = ctrl.prometheusUrl.Scheme
+		req.URL.Host = ctrl.prometheusUrl.Host
+		req.URL.Path = c.Param("proxyPath")
+	}
```

Comment:
I don't like the fact, that the proxy is initialized with every incoming request. what about using the controller with composition?

```
type Controller struct {
	prometheusUrl *url.URL
	httputil.ReverseProxy
}
```

for c.Param("proxyPath") you could use the context to pass this value

Reply:
You are completely right. This could be done much more easily :)